### ISSUE TITLE
Exception: Unable to execute HTTP request: Channel was closed before it could be written to

### ISSUE THREAD

**Issue by lorenariel** [aws/aws-sdk-java-v2#1766]
## Describe the issue

I’m developing docker spring boot application service which publish a data to aws S3.
I’m using the S3AsyncClient with PutObjectRequest api.
The service is deployed both on aws ECS and swarm, but getting always the following exception:
 java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Channel was closed before it could be written to

It’s important to mention that when running the same program on my local laptop I don’t have any issue when using S3AsyncClient or S3Client.

Also, I tested the S3Client With docker service on ECS and swarm and don’t have any issue, managed to put object on the aws bucket, the problem is only with the S3AsyncClient.

I reviewed the exception in many blogs and found that the exception can be related to connectionMaxIdleTime or to increase the retry from the default.
I set the connectionMaxIdleTime = 5 and the retry to 5 and still have the issue on ECS or swarm…. For S3AsyncClient and not for S3Client.

## Steps to Reproduce

1. _**Source code for S3AsyncClient:**_

       ClientOverrideConfiguration overrideConfiguration =
                ClientOverrideConfiguration.builder()
                        .apiCallTimeout(Duration.ofSeconds(30))
                        .apiCallAttemptTimeout(Duration.ofSeconds(20))
                        .retryPolicy(RetryPolicy.builder()
                                .numRetries(5)
                                .retryCondition(RetryCondition.defaultRetryCondition())
                                .build()
                        )
                        .build();

        SdkAsyncHttpClient httpClient = NettyNioAsyncHttpClient.builder()
                .connectionTimeout(Duration.ofSeconds(20))
                .connectionAcquisitionTimeout(Duration.ofSeconds(20))
                .connectionMaxIdleTime(Duration.ofSeconds(5))
                .build();

        S3AsyncClient s3 = S3AsyncClient.builder()
                .region(Region.US_EAST_1)
                .credentialsProvider(c)
                .overrideConfiguration(overrideConfiguration)
                .httpClient(httpClient)
                .asyncConfiguration(
                        b -> b.advancedOption(SdkAdvancedAsyncClientOption
                                        .FUTURE_COMPLETION_EXECUTOR,
                                Executors.newFixedThreadPool(2)
                        )
                )
                .build();

        try {
            for (int idx = 0; idx < 1; idx++) {
                log.debug("------------before future-----------");

                CompletableFuture<PutObjectResponse> f = s3.putObject(PutObjectRequest.builder()
                        .bucket("caas-msg-dev-1").key("oren/test")
                        .build(), AsyncRequestBody.empty());

                f.whenComplete((resp, err) -> {
                    if (resp != null) {
                        log.debug("S3 response: " + resp);
                    } else {
                        log.error(err.getMessage());
                    }

                });

                f.join();
                log.debug("------------after future-----------");
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        } finally {
            log.debug("------------closing s3 client-----------");
            s3.close();
        }

1. **_Source code for S3Client:_**

        S3Client s3 = S3Client.builder()
                .region(Region.US_EAST_1)
                .credentialsProvider(c)
                .build();

        log.debug("--- before -----");
        s3.putObject(PutObjectRequest.builder()
                .bucket("caas-msg-dev-1").key("oren/test")
                .build(), RequestBody.empty());

        s3.listBuckets().buckets().forEach(v -> log.debug("--->" + v));
        log.debug("--- after -----");

## Your Environment

* AWS Java SDK version used:
compile  group: 'software.amazon.awssdk', name: 's3', version: '2.11.6'
compile 'software.amazon.awssdk:netty-nio-client:2.11.6'

* JDK version used: 
java 8
'org.springframework.boot' version '2.2.6.RELEASE'

* Operating System and version: docker

**Comment by lorenariel** [comment]
After adding more debug level:

Allow unsafe renegotiation: false,
Allow legacy hello messages: true,
 initial handshake: true,
Is secure renegotiation: false,
No available cipher suite for TLSv1,
No available cipher suite for TLSv1.1,
No available cipher suite for TLSv1.2,
aws-java-sdk-NettyEventLoop-0-2, fatal error: 40: Couldn't kickstart handshaking,
javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate),
aws-java-sdk-NettyEventLoop-0-2, SEND TLSv1.2 ALERT:  fatal, description = handshake_failure,
aws-java-sdk-NettyEventLoop-0-2, WRITE: TLSv1.2 Alert, length = 2,
aws-java-sdk-NettyEventLoop-0-2, called closeOutbound(),
aws-java-sdk-NettyEventLoop-0-2, closeOutboundInternal(),
aws-java-sdk-NettyEventLoop-0-2, called closeInbound(),
aws-java-sdk-NettyEventLoop-0-2, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?,
Using SSLEngineImpl.,
2020-04-05 15:12:10.622 DEBUG --- [tyEventLoop-0-4] [                              ] io.netty.handler.logging.LoggingHandler  : [id: 0xb909d05e] REGISTERED,
2020-04-05 15:12:10.622 DEBUG --- [tyEventLoop-0-4] [                              ] io.netty.handler.logging.LoggingHandler  : [id: 0xb909d05e] CONNECT: *******,
2020-04-05 15:12:10.623 DEBUG --- [tyEventLoop-0-2] [                              ] io.netty.handler.logging.LoggingHandler  : [id: 0x4b3c8639, L:/****] USER_EVENT: SslHandshakeCompletionEvent(javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)),
Allow unsafe renegotiation: false,
Allow legacy hello messages: true,
 initial handshake: true,
Is secure renegotiation: false,
No available cipher suite for TLSv1,
No available cipher suite for TLSv1.1,
No available cipher suite for TLSv1.2,
aws-java-sdk-NettyEventLoop-0-4, fatal error: 40: Couldn't kickstart handshaking,
javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate),
aws-java-sdk-NettyEventLoop-0-4, SEND TLSv1.2 ALERT:  fatal, description = handshake_failure,
aws-java-sdk-NettyEventLoop-0-4, WRITE: TLSv1.2 Alert, length = 2,
aws-java-sdk-NettyEventLoop-0-4, called closeOutbound(),
aws-java-sdk-NettyEventLoop-0-4, closeOutboundInternal(),
aws-java-sdk-NettyEventLoop-0-4, called closeInbound(),

**Comment by zoewangg** [comment]
@lorenariel From the logs, it seems the root cause is that there was no available cipher suite for TLSv1, TLSv1.1 or TLSv1.2 on the docker machine. 

Could you add the following statements to see what cipher suites are supported on the machine?

```
System.out.println(Arrays.toString(SSLContext.getDefault().getSupportedSSLParameters().getProtocols()));

System.out.println(Arrays.toString(SSLContext.getDefault().getSupportedSSLParameters().getCipherSuites()));
```

**Comment by debora-ito** [comment]
Closing this due to inactivity. Feel free to reopen if you have further questions.

**Comment by mistreckless** [comment]
Please reopen this issue, it still actual. I've got the same error with `CloudWatchLogsAsyncClient.` Without docker everything works fine but when I try to run my app inside a docker container I got 
`software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Channel was closed before it could be written to.`

Inside docker container
- `System.out.println(Arrays.toString(SSLContext.getDefault().getSupportedSSLParameters().getProtocols()));` prints `[TLSv1, TLSv1.1, TLSv1.2]`
- `System.out.println(Arrays.toString(SSLContext.getDefault().getSupportedSSLParameters().getCipherSuites()));` prints `[TLS_EMPTY_RENEGOTIATION_INFO_SCSV, SSL_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384, SSL_ECDHE_RSA_WITH_AES_256_CBC_SHA384, SSL_RSA_WITH_AES_256_CBC_SHA256, SSL_ECDH_ECDSA_WITH_AES_256_CBC_SHA384, SSL_ECDH_RSA_WITH_AES_256_CBC_SHA384, SSL_DHE_RSA_WITH_AES_256_CBC_SHA256, SSL_DHE_DSS_WITH_AES_256_CBC_SHA256, SSL_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, SSL_ECDHE_RSA_WITH_AES_256_CBC_SHA, SSL_RSA_WITH_AES_256_CBC_SHA, SSL_ECDH_ECDSA_WITH_AES_256_CBC_SHA, SSL_ECDH_RSA_WITH_AES_256_CBC_SHA, SSL_DHE_RSA_WITH_AES_256_CBC_SHA, SSL_DHE_DSS_WITH_AES_256_CBC_SHA, SSL_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,  SSL_ECDHE_RSA_WITH_AES_128_CBC_SHA256, SSL_RSA_WITH_AES_128_CBC_SHA256, SSL_ECDH_ECDSA_WITH_AES_128_CBC_SHA256, SSL_ECDH_RSA_WITH_AES_128_CBC_SHA256, SSL_DHE_RSA_WITH_AES_128_CBC_SHA256, SSL_DHE_DSS_WITH_AES_128_CBC_SHA256, SSL_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, SSL_ECDHE_RSA_WITH_AES_128_CBC_SHA, SSL_RSA_WITH_AES_128_CBC_SHA, SSL_ECDH_ECDSA_WITH_AES_128_CBC_SHA, SSL_ECDH_RSA_WITH_AES_128_CBC_SHA, SSL_DHE_RSA_WITH_AES_128_CBC_SHA, SSL_DHE_DSS_WITH_AES_128_CBC_SHA, SSL_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, SSL_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, SSL_ECDHE_RSA_WITH_AES_256_GCM_SHA384, SSL_RSA_WITH_AES_256_GCM_SHA384, SSL_ECDH_ECDSA_WITH_AES_256_GCM_SHA384, SSL_ECDH_RSA_WITH_AES_256_GCM_SHA384, SSL_DHE_DSS_WITH_AES_256_GCM_SHA384, SSL_DHE_RSA_WITH_AES_256_GCM_SHA384, SSL_ECDHE_RSA_WITH_AES_128_GCM_SHA256, SSL_RSA_WITH_AES_128_GCM_SHA256, SSL_ECDH_ECDSA_WITH_AES_128_GCM_SHA256, SSL_ECDH_RSA_WITH_AES_128_GCM_SHA256, SSL_DHE_RSA_WITH_AES_128_GCM_SHA256, SSL_DHE_DSS_WITH_AES_128_GCM_SHA256, SSL_DH_anon_WITH_AES_256_GCM_SHA384, SSL_DH_anon_WITH_AES_128_GCM_SHA256, SSL_DH_anon_WITH_AES_256_CBC_SHA256, SSL_ECDH_anon_WITH_AES_256_CBC_SHA, SSL_DH_anon_WITH_AES_256_CBC_SHA, SSL_DH_anon_WITH_AES_128_CBC_SHA256, SSL_ECDH_anon_WITH_AES_128_CBC_SHA, SSL_DH_anon_WITH_AES_128_CBC_SHA, SSL_RSA_WITH_NULL_SHA256, SSL_ECDHE_ECDSA_WITH_NULL_SHA, SSL_ECDHE_RSA_WITH_NULL_SHA, SSL_RSA_WITH_NULL_SHA, SSL_ECDH_ECDSA_WITH_NULL_SHA, SSL_ECDH_RSA_WITH_NULL_SHA, SSL_ECDH_anon_WITH_NULL_SHA, SSL_RSA_WITH_NULL_MD5]`

I tried to disable tls, but nothing changes. Here my peace of code
```
    @Bean
    open fun cloudWatchLoggerClient(): CloudWatchLogsAsyncClient {
        println(Arrays.toString(SSLContext.getDefault().supportedSSLParameters.protocols))
        println(Arrays.toString(SSLContext.getDefault().supportedSSLParameters.cipherSuites))
        return CloudWatchLogsAsyncClient.builder()
            .credentialsProvider {
                AwsBasicCredentials.create(awsCredentials.accessKey, awsCredentials.secretKey)
            }
            .httpClient(NettyNioAsyncHttpClient.builder()
                .buildWithDefaults(AttributeMap.builder()
                    .put(SdkHttpConfigurationOption.TLS_KEY_MANAGERS_PROVIDER, TlsKeyManagersProvider.noneProvider())
                    .put(SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES, true)
                    .build()))
            .region(Region.of(loggerProperties.regionId))
            .build()
    }
```

It really hard to debug exceptions which happen only in docker container. Here the stack trace of the exception
```
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Channel was closed before it could be written to.
        at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:43) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException(RetryableStageHelper.java:205) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException(RetryableStageHelper.java:201) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:143) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:125) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor$$Lambda$667.00000000BD969390.accept(Unknown Source) ~[na:na]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:771) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:747) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:485) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1988) ~[na:1.8.0]
        at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:74) ~[utils-2.14.12.jar!/:na]
        at software.amazon.awssdk.utils.CompletableFutureUtils$$Lambda$661.00000000BCC58B50.accept(Unknown Source) ~[na:na]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:771) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:747) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:485) ~[na:1.8.0]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1988) ~[na:1.8.0]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$WrappedErrorForwardingResponseHandler.onError(MakeAsyncHttpRequestStage.java:126) ~[sdk-core-2.14.12.jar!/:na]
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.handleFailure(NettyRequestExecutor.java:299) ~[netty-nio-client-2.14.12.jar!/:na]
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.tryConfigurePipeline(NettyRequestExecutor.java:219) ~[netty-nio-client-2.14.12.jar!/:na]
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.makeRequestListener(NettyRequestExecutor.java:167) ~[netty-nio-client-2.14.12.jar!/:na]
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor$$Lambda$659.00000000BCC581F0.operationComplete(Unknown Source) ~[na:na]
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.DefaultPromise.access$200(DefaultPromise.java:35) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:501) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) ~[netty-transport-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.48.Final.jar!/:4.1.48.Final]
        at java.lang.Thread.run(Thread.java:812) ~[na:2.9 (11-06-2018)]
```

**Comment by unoexperto** [comment]
Same issue with 2.16.95. I'm on Oracle JDK 16 under Ubuntu 21.04. Any suggestions, folks ?

@debora-ito

**Comment by Bennett-Lynch** [comment]
Hi @lorenariel. A change was released today as part of version `2.17.101` that should help significantly reduce the occurrence of `Channel was closed before it could be written to` errors. If you are still experiencing this issue, please try the latest version and let us know if you see improvement. Change details can be found in https://github.com/aws/aws-sdk-java-v2/pull/2883.
